### DISCUSSION TITLE
Am I not using route-based IPSec?

### DISCUSSION THREAD

**Discussion by t-dubbb** [strongswan/strongswan#2334] [Q&A]
Hello, I am running Strongswan 5.9.12 on Alpine Linux 3.17, kernel 5.15.162-LTS. I am going back and forth with Cisco TAC on an issue I am having with one of their routers. I have a site-to-site IKEv2 between Strongswan and Cisco. The problem is that the Cisco is establishing additional, duplicate child SA tunnels every 30 seconds. This leads to a massive accumulation of identical tunnels. According to Cisco, I am not using route-based configuration on Strongswan, and this is the source of the problem. I really tried to follow the official page on the Strongswan web site on setting up a route based tunnel using XFRM interfaces, but I guess I am missing something? To be honest there are many pages on the Strongswan web site that don't seem clear or complete to me. I am posting my swanctl.conf config below. Before Strongswan starts I am creating the XFRM interface using the commands "/sbin/ip link add ipsec0 type xfrm dev eth4 if_id 0x1" and "/sbin/ip link set dev ipsec0 up". The tunnel is up and traffic appears to be flowing normally without interruption, and doing a tcpdump on the ipsec0 interface shows all expected traffic. Let me know what I am doing wrong, thanks in advance.

	TUNNEL {
		remote_addrs = MYTUNNEL.COM
		version = 2
		proposals = aes128-sha256-modp2048
		keyingtries = 0
		dpd_delay = 300s
		dpd_timeout = 1500s
		if_id_in = 1
		if_id_out = 1
		LOCAL {
			auth = pubkey
			certs = MYCERT.crt
		}
		REMOTE {
			auth = pubkey
			id = “CN=REMOTE_CERT_DN”
		}
		children {
			TUNNEL {
				local_ts = 172.16.16.0/22
				remote_ts = 192.168.192.0/22
				esp_proposals = aes128-sha256-modp2048
				rekey_time = 1h
				dpd_action = restart
				start_action = trap|start
				set_mark_out = 1
			}
		}
	}

**Comment by tobiasbrunner** [comment]
> According to Cisco, I am not using route-based configuration on Strongswan, and this is the source of the problem.

I depends a bit on the exact approach, but with XFRM interfaces that's basically an entirely local decision and should not affect the other end (it's not something that's negotiated or indicated to the peer). In particular, there should be absolutely no reason to create additional CHILD_SAs if the first one's traffic selectors already cover all the traffic. Your claim that duplicate CHILD_SAs are created indicates that the peer creates additional CHIILD_SA with the same traffic selectors (or a subset thereof, maybe for individual IPs?), which makes no sense and using a route-based approach should have no impact on in whatsoever. However, it's possible that the Cisco guys understand something different under the term and that there is some configuration mismatch. So you have to check with them what exactly it is they expect to be different.

**Comment by t-dubbb** [comment]
It's not a "claim" that there are duplicate SAs, it's been verified by multiple Cisco TAC agents as well as myself.
Cisco  are  saying that I should be using 0.0.0.0/0 for the local/remote subnets on the Strongswan side. What I don't understand is that if the subnets are 0.0.0.0/0, how will Strongswan know the actual remote subnet routes to automatically install? Indeed, I tried changing the subnets to 0.0.0.0/0 and Strongswan installed a 0.0.0.0/0 route to the XFRM interface, which (predictably) broke internet connectivity entirely for the machine it is running on, since everything was now trying to route to the XFRM. At the moment, I have solved the issue by giving up on having Strongswan automatically install routes, configuring Strongswan with 0.0.0.0/0 for local/remote subnets, disabling the charon.plugins.kernel-netlink.install_routes_xfrmi plugin, and manually routing the specific remote subnet to the XFRM interface. This is fine I suppose, but I don't really understand how the "automatically install routes" feature is supposed to work.

**Reply by tobiasbrunner** [reply]
> It's not a "claim" that there are duplicate SAs, it's been verified by multiple Cisco TAC agents as well as myself.

But you haven't provided any logs, status output etc., so all we have here is your claim.



> Cisco are saying that I should be using 0.0.0.0/0 for the local/remote subnets on the Strongswan side.

So why don't you?

> What I don't understand is that if the subnets are 0.0.0.0/0, how will Strongswan know the actual remote subnet routes to automatically install?

It doesn't. That's often the whole point of using route-based IPsec, the traffic that's protected is flexibly controlled via routing.

> disabling the charon.plugins.kernel-netlink.install_routes_xfrmi plugin

Which is not enabled by default.

> This is fine I suppose, but I don't really understand how the "automatically install routes" feature is supposed to work.

Obviously, not intended for the `0.0.0.0/0` scenario (except for roadwarriors that actually want to tunnel everything to the server). And if the traffic selectors overlap with the IKE traffic, some care is necessary to avoid  a routing loop (but that's true whether strongSwan installs the routes or not).

**Reply by t-dubbb** [reply]
It's late at night where I am, and I don't have the time right now to sanitize logs and otherwise respond to your statements. I am going to say that you have an extremely condescending attitude. Time and again, for many years, I have seen this kind of antagonistic interaction from you on various mailing lists and message boards, frequently to people who have done nothing but ask questions. It's an interaction style that is greatly off-putting, and it is no surprise when the majority of your replies like this get no response. I'll try to give a detailed to reply to the things you said in the future. Or I may not, who knows. Sheesh.......
